src/pybind/mgr/telemetry/module.py

```diff
@@ -939,6 +1036,117 @@ class EndPoint(enum.Enum):
         ceph = 'ceph'
         device = 'device'
 
+    def collection_delta(self) -> List[Collection]:
+        '''
+        Find collections that are available in the module, but are not in the db
+        '''
+        new_collection : List[Collection] = []
+
+        for c in MODULE_COLLECTION:
+            for k, v in c.items():
+                if k == 'name' and v.name not in self.db_collection:
+                    new_collection.append(v)
+
+        return new_collection
+    
+    def is_major_upgrade(self) -> bool:
+        '''
+        Returns True only if the user last opted-in to an older major
+        '''
+        if self.last_opted_in_ceph_version is None or self.last_opted_in_ceph_version == 0:
+            # we do not know what Ceph version was when the user last opted-in,
+            # thus we do not wish to nag in case of a major upgrade
+            return False
+
+        mon_map = self.get('mon_map')
+        mon_min = mon_map.get("min_mon_release", 0)
+
+        if mon_min - self.last_opted_in_ceph_version > 0:
+            self.log.debug(f"major upgrade: mon_min is is: {mon_min} and user last opted-in in {self.last_opted_in_ceph_version}")
+            return True
+
+        return False
+
+    def is_opted_in(self) -> bool:
+        # If len is 0 it means that the user is either opted-out (never
+        # opted-in, or invoked `telemetry off`), or they upgraded from a
+        # telemetry revision 1 or 2, which required to re-opt in to revision 3,
+        # regardless, hence is considered as opted-out
+        return len(self.db_collection) > 0
+
+    def should_nag(self) -> bool:
+        # Find delta between opted-in collections and module collections;
+        # nag only if module has a collection which is not in db, and nag == True.
+
+        # We currently do not nag if the user is opted-out (or never opted-in).
+        # If we wish to do this in the future, we need to have a tri-mode state
+        # (opted in, opted out, no action yet), and it needs to be guarded by a
+        # config option (so that nagging can be turned off via config).
+        # We also need to add a last_opted_out_ceph_version variable, for the
+        # major upgrade check.
+
+        # check if there are collections the user is not opt-in to
+        # that we should nag about
+        for c in MODULE_COLLECTION:
+            for k, v in c.items():
+                if k == 'name' and v.name not in self.db_collection:
+                    if c['nag'] == True:
+                        self.log.debug(f"The collection: {v} is not reported, and we should nag about it")
+                        return True
+
+        # user might be opted-in to the most recent collection, or there is no
+        # new collection which requires nagging about
+        return self.is_major_upgrade()
```

Comment:
could there be a major upgrade with no new collections at all, that we don't need to nag about?